src/rtc.c

```diff
@@ -346,17 +345,41 @@ void RTC_init(hourFormat_t format, sourceClock_t source, bool reset)
   RtcHandle.Init.OutPutType = RTC_OUTPUT_TYPE_OPENDRAIN;
 #endif /* STM32F1xx */
 
+  /* Ensure backup domain is enabled before we init the RTC so we can use the backup registers for date retention on stm32f1xx baords */
+  enableBackupDomain();
+
   HAL_RTC_Init(&RtcHandle);
 
+#if defined(STM32F1xx)
+  // Copy date data back out of the BackUp registers
+  RTC_DateTypeDef BackupDate;
+  RTC_TimeTypeDef DummyTime;
+  uint32_t dateMem;
+  dateMem = HAL_RTCEx_BKUPRead(&RtcHandle, RTC_BKP_DR3);
+  dateMem |= HAL_RTCEx_BKUPRead(&RtcHandle, RTC_BKP_DR2) << 16;
+  memcpy(&BackupDate, &dateMem, sizeof(uint32_t));
+  if (IS_RTC_YEAR(BackupDate.Year) && IS_RTC_MONTH(BackupDate.Month) && IS_RTC_DATE(BackupDate.Date)) {
+    /* Change the date retrieved from the backup registers */
+    RtcHandle.DateToUpdate.Year  = BackupDate.Year;
+    RtcHandle.DateToUpdate.Month = BackupDate.Month;
+    RtcHandle.DateToUpdate.Date  = BackupDate.Date;
```

Comment:
One thing is still missing: weekday. After power cycle weekday resets to 0 (Sunday). Simple code and steps to reproduce:
```
STM32RTC& rtc = STM32RTC::getInstance();
rtc.setClockSource(STM32RTC::LSE_CLOCK);
rtc.begin(false,STM32RTC::HOUR_24);
if(!rtc.isTimeSet())
{
    rtc.setDate(22,2,21);
}
println(rtc.getWeekDay());
```
1. First run (VDD and VBAT not present before, so RTC not runs): result is 1 (Monday which is correct for 22-02-2021)
2. Power cycle with VBAT supplied
3. Second run after power cycle: result is 0 (Sunday), date restored correctly, but without a weekday.

Crude workaround -> add these 3 lines after `rtc.begin()`:
```
uint8_t wd, d, m, y;
rtc.getDate(&wd,&d,&m,&y);
rtc.setDate(d,m,y);
```
It works because function `setDate()` leads to invoking `RTC_WeekDayNum()` from `stm32f1xx_hal_rtc.c` which recalculates weekday for given current date.

From a quick code analysis I think the problem have simple solution, just add:
`RtcHandle.DateToUpdate.WeekDay = BackupDate.WeekDay;`

Reply:
I have re implemented the weekday retention, I remember testing it in my development environment but I must have removed it when I rolled back some changes.

Reply:
Ok, tested, works great. But I've found more things that could be improved a little bit:

This line https://github.com/josh-gaby/STM32RTC/blob/c2d233f9012e5d66fa21798d47781d54e562832d/src/rtc.c#L361 should be extended with weekday check:
```
if (IS_RTC_YEAR(BackupDate.Year) && IS_RTC_MONTH(BackupDate.Month) && IS_RTC_DATE(BackupDate.Date) && IS_RTC_WEEKDAY(BackupDate.WeekDay))
```
Same validation is used in `RTC_SetDate()`.

And if we're talking about `IS_RTC_WEEKDAY` define (in `rtc.h`), and the `#if` block surrounding it - both could be done better:
 1. Construction with 7 OR'ed equality checks is imho weird, should be done like `IS_RTC_MONTH` (from `stm32f1xx_hal_rtc.h`)
 2. Inside `#ifdef` block with `IS_RTC_WEEKDAY` there is also a re-define for `IS_RTC_HOUR12`. I think it should be moved from here because `!defined(IS_RTC_WEEKDAY)` have nothing to do with 12/24 format.
 
My proposition for the whole section:
```
#if defined(STM32F1xx) && !defined(IS_RTC_WEEKDAY)
/* Compensate missing HAL definition */
#define IS_RTC_WEEKDAY(WEEKDAY) (((WEEKDAY) >= 0U) && ((WEEKDAY)) <= 6U)) 
#endif /* !STM32F1xx && !IS_RTC_WEEKDAY */

#if defined(STM32F1xx)
/* F1 doesn't manage 12h format */
#define IS_RTC_HOUR12(HOUR)      IS_RTC_HOUR24(HOUR)
#endif /* !STM32F1xx */
```
And yes, I know above changes are only a little bit related to main goal of PR, but why not improve these things by the way ;)

Reply:
That validation is used in the `RTC_SetDate` function because it is dealing with potentially "user" input where as this is only dealing with data that was saved from known good values, however I have added a check for valid weekday but only on the weekday itself, on the basis that if somehow only the weekday was mangled by the backup registers (not likely) then I would rather the date still be set.

I'm not going to be changing those defines though as that's outside the scope of this pull request, I suggest creating a separate pull request if you want them changed.

Reply:
Ok, looks good to me. And You're right about defines, I'll follow Your suggestion and make a new PR.